### PR TITLE
feat(core): add support for fallback content in ng-content

### LINKED ISSUE #12530: ng-content default content
**I'm submitting a ...**  (check one with "x")

```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When you add something inside a ng-content tag you get an error message: <ng-content> element cannot have content.

**Expected behavior**
Should render the html inside the ng-content tag as a default if no content was given. 

**Minimal reproduction of the problem with instructions**
Create a component that uses ng-content somewhere. Put some html inside, that should be rendered when nothing else is 'injected' there from outside. Then use the component somewhere else. 

**What is the motivation / use case for changing the behavior?**
In some cases you need a default behavior that can be replaced. In my case its an app header that shows the logo. When you navigate to a sub-page the logo will be replaced by a back button. It feels strange to add something like an input property to hide or show the logo. Some default content for ng content placeholders would feel much more intuitive.

**Please tell us about your environment:**
Win10, Angular CLI beta 18, VS Code
- **Angular version:** 2.1.0
  2.1.0
- **Browser:** 
  All
- **Language:**
  TS
- **Node (for AoT issues):** `node --version` =

### PR DESCRIPTION
**Pull request by crisbeto**
Adds the ability to specify content that Angular should fall back to if nothing is projected into an `ng-content` slot. For example, if we have the following setup

```
@Component({
  selector: 'my-comp',
  template: `
    <ng-content select="header">Default header</ng-content>
    <ng-content select="footer">Default footer</ng-content>
  `
})
class MyComp {}

@Component({
  template: `
    <my-comp>
      <footer>New footer</footer>
    </my-comp>
  `
})
class MyApp {}
```

The instance of `my-comp` in the app will have the default header and the new footer.

**Note:** Angular's content projection happens during creation time. This means that dynamically changing the contents of the slot will not cause the default content to show up, e.g. if a `if` block goes from `true` to `false`.

Fixes #12530.

### CODE REVIEW

File: packages/core/test/acceptance/content_spec.ts [hunk 1]
```diff
@@ -384,12 +384,12 @@ describe('projection', () => {
     @Component({
       selector: 'child',
       template:
-          `<div *ngFor="let item of [1, 2]; let i = index">({{i}}):<ng-content></ng-content></div>`
+          '<div *ngFor="let item of [1, 2]; let i = index">({{i}}):<ng-content></ng-content></div>'
```
Comment (crisbeto) [inline packages/core/test/acceptance/content_spec.ts:387]: This change is unrelated, it was just messing up the syntax highlighting for the rest of the file for some reason.

File: packages/core/src/render3/instructions/projection.ts [hunk 2]
```diff
@@ -109,31 +115,66 @@ export function ɵɵprojectionDef(projectionSlots?: ProjectionSlots): void {
...
 
-  // `<ng-content>` has no content
+  // `<ng-content>` has no content. Even if there's fallback
+  // content, the fallback is shown next to it.
   setCurrentTNodeAsNotParent();
 
   const hydrationInfo = lView[HYDRATION];
   const isNodeCreationMode = !hydrationInfo || isInSkipHydrationBlock();
-  if (isNodeCreationMode &&
+  const componentNode = lView[DECLARATION_COMPONENT_VIEW][T_HOST] as TElementNode;
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions/projection.ts:150]: super nit: how about `componentHostNode` or sth else that indicates a host?

File: packages/core/test/acceptance/content_spec.ts [hunk 3]
```diff
@@ -1368,4 +1368,400 @@ describe('projection', () => {
...
+      }
+
+      const fixture = TestBed.createComponent(App);
+      expect(getElementHtml(fixture.nativeElement))
+          .toContain(
+              `<projection><span one="">One</span>|` +
+              `Catch-all fallback|Two fallback<div three="">Three</div></projection>`);
+    });
+
+    it('should render the catch-all slots fallback content if the element only contains comments',
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/content_spec.ts:1407]: Great test, thank you for thinking about this one!

File: packages/core/test/acceptance/content_spec.ts [hunk 4]
```diff
@@ -1368,4 +1368,400 @@ describe('projection', () => {
...
+
+      const fixture = TestBed.createComponent(App);
+      fixture.detectChanges();
+      expect(getElementHtml(fixture.nativeElement)).toContain(`<projection></projection>`);
+
+      fixture.componentInstance.projection.createContent();
+      fixture.detectChanges();
+      expect(getElementHtml(fixture.nativeElement)).toContain(`<projection>Fallback</projection>`);
+    });
+  });
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/content_spec.ts:1766]: Excellent tests @crisbeto! The only other cases I could think of are around re-projection. We should add a test verifying proper behaviour of the projection node with default content being re-projected (with the default content I presume?)

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM overall, thank you for adding very thorough tests. The only addition I could think of are tests for re-projection with the default content.

**Review by atscott — APPROVED** [review]
reviewed-for: language-service

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 2fc11eae9ea65160866bf7ba46c10520ae9a141f.
